### PR TITLE
Fix stubbed String constructor special case

### PR DESCRIPTION
**Pull request by antlechner**
This PR fixes a bug that was caused/uncovered by https://github.com/diffblue/cbmc/pull/4483.
Generally JBMC should only be run on methods that include Strings if a model for String is loaded. However, `--show-properties` should work even without the model. After the changes in https://github.com/diffblue/cbmc/pull/4483, in the special case where the Java method includes a call to a String constructor and no model for it was loaded, we didn't check if the object had a string type, and a pointer to an unsignedbv ended up in the pointer case of the object factory, causing an invariant violation. This PR fixes the conditions in the struct case of the object factory so it always checks if the type is a String type, not only when `skip_classid` is false.

- [x] Each commit message has a non-empty body, explaining why the change was made. [except for the new test, which has a description in its .desc file]
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- n/a Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: jbmc/src/java_bytecode/java_object_factory.cpp [hunk 1]
```diff
@@ -794,77 +794,69 @@ void java_object_factoryt::gen_nondet_struct_init(
   // * Not if the object has already been initialised by our caller, in which
   //   case they will set `skip_classid`
   // * Not if we're re-initializing an existing object (i.e. update_in_place)
-
-  bool skip_special_string_fields = false;
-
-  if(!is_sub &&
-     !skip_classid &&
-     update_in_place != update_in_placet::MUST_UPDATE_IN_PLACE)
+  // * Always if it has a string type (special case).
```
Comment (allredj) [inline jbmc/src/java_bytecode/java_object_factory.cpp]: :pick: I don't think `(special case)` is necessary here. Rather an explanation of *why* strings are a special case would be useful.

File: jbmc/src/java_bytecode/java_object_factory.cpp [hunk 2]
```diff
@@ -794,77 +794,69 @@ void java_object_factoryt::gen_nondet_struct_init(
...
+
+  const bool is_char_sequence =
+    java_string_library_preprocesst ::implements_java_char_sequence(
```
Comment (allredj) [inline jbmc/src/java_bytecode/java_object_factory.cpp]: :pick: :question:  Is it customary to have a space before the double colon?

File: jbmc/src/java_bytecode/java_object_factory.cpp [hunk 3]
```diff
@@ -801,12 +801,11 @@ void java_object_factoryt::gen_nondet_struct_init(
       struct_type);
   const bool has_length_and_data =
     struct_type.has_component("length") && struct_type.has_component("data");
-  const bool skip_special_string_fields =
-    is_char_sequence && has_length_and_data;
+  const bool is_string_type = is_char_sequence && has_length_and_data;
```
Comment (allredj) [inline jbmc/src/java_bytecode/java_object_factory.cpp:806]: :+1:  Thanks for changing that. The name was really bad.

**Review by allredj — APPROVED** [review]

File: jbmc/src/java_bytecode/java_object_factory.cpp (hunk 2 above)
Comment (smowton) [reply jbmc/src/java_bytecode/java_object_factory.cpp]: No :p

**Review by smowton — APPROVED** [review]

**Review by thomasspriggs — APPROVED** [review]

File: jbmc/src/java_bytecode/java_object_factory.cpp (hunk 2 above)
Comment (antlechner) [reply jbmc/src/java_bytecode/java_object_factory.cpp]: Huh, clang-format did this, not sure why! I'll fix it as CI is failing with a linting error anyway.

File: jbmc/src/java_bytecode/java_object_factory.cpp (hunk 1 above)
Comment (antlechner) [reply jbmc/src/java_bytecode/java_object_factory.cpp]: I added a few more lines, with a reference to the function that initialises string fields. Any further documentation should probably go to that function.

**Review by allredj — COMMENTED** [review]
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 9bc1a6ed47b57c5a8211d16e8d76367a2c89e16b).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/107467585
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

**Review by allredj — COMMENTED** [review]
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: c6e55522b3e7b057716a74f59f8956d2b91d7323).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/107467525
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

**Pull request merged**